internal/controller/pkg/revision/reconciler.go

```diff
@@ -468,6 +469,20 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 		return reconcile.Result{Requeue: false}, nil
 	}
 
+	// Check the pause annotation and return if it has the value "true"
+	// after logging, publishing an event and updating the SYNC status condition
+	if meta.IsPaused(pr) {
```

Comment:
We should move this up before `meta.wasDeleted` check so that the pause will be effective for package revisins being deleted. This also aligns with how the feature was implemented in the managed reconciler and claim reconciler.

Reply:
Meaning that we cannot remove paused objects?

Reply:
Yes, deleted objects would still need reconciliation to finalize their deletion. 
One can remove a paused object after removing the paused annotation.

Reply:
Is that really what users expect to happen? In [the docs](https://docs.crossplane.io/latest/concepts/managed-resources/#paused) we do not say anything about being impossible to remove paused objects.

Imagine the following situation: a user attempts to remove a namespace containing a paused claim. This operation would hangs forever, right?

Reply:
> Is that really what users expect to happen?

As said, deletion is also part of reconciliation, the controller needs some operation / clean-up to finalize deletion. I would be surprised that part of reconciliation was not "paused". And that could be more desirable if I want to debug before the resource is deleted.

Another point is, one can go and find why their resource not deleted if that is the behavior they are looking for and remove the annotation to proceed. But if we do the opposite, there wouldn't be a way to revert for a user expecting the deletion to hang on with paused annotation, as the resource would already be gone.

> Imagine the following situation: a user attempts to remove a namespace containing a paused claim. This operation would hangs forever, right?

They can see why it hangs and remove the paused annotation to proceed. 
I don't think this is a feature that people would end up having paused resources around. It is more like a deliberate action where people would revert after they get what they are after.

Reply:
I agree with @turkenh, pausing should pause everything, it would be surprising otherwise.

Reply:
I tend to be on @pedjak's side here. Broken deletion seems to be unexpected and strange.

Reply:
Just tried the following:
* created namespace `example`
* created claim in it
* paused claim
* tried to remove namespace with `kubectl delete ns example`
* `kubectl` hangs with the message `namespace "example" deleted` but it does not return to the prompt

How a user can know that there is a claim in the namespace holding the delete?

Even if I do `kubectl describe` on that claim, there is no trace saying that deletion is blocked and can be resumed when pause annotation is removed.

Users do not need to know anything about how things are working under the hood, how many controllers are processing objects, in what order, etc. However, when users want to remove an object, we should either protect against that with a proper error message, or allow him to do that. If an object reconciliation is paused, we can still allow removal, cleaning up what is needed, etc.

Reply:
Let me link the discussion on slack: https://crossplane.slack.com/archives/CEF5N8X08/p1697618937115119?thread_ts=1697555426.091559&cid=CEF5N8X08

> Even if I do `kubectl describe` on that claim, there is no trace saying that deletion is blocked and can be resumed when pause annotation is removed.

This is something different. There should be. Shouldn't this output some events/logs: https://github.com/crossplane/crossplane/blob/af2496ebacd6227f19be3b1fb8af58100a4d8b2e/internal/controller/apiextensions/claim/reconciler.go#L371-L379

As I stated in the thread, it feels like we are trying to optimize user experience for a very specific scenario while ignoring others who would expect the reconciliation to be paused, including the deletion.

FWIW, it is also implemented in the same way in cluster API, which could be considered as an indication that this is what most people expect from such a functionality, see: [https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/b90810b70864664a00314f40a3f8a31d4215f1b5/controllers/awscluster_controller.go#L[…]4](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/b90810b70864664a00314f40a3f8a31d4215f1b5/controllers/awscluster_controller.go#L165-L194)

Reply:
> This is something different. There should be. Shouldn't this output some events/logs: 

The events are gone, because when you try to remove a namespace, k8s starts removing all of them in the given namespace, including events. Given that there is no problems to remove them (no finalizers, no blockers), the events are all gone.

> it feels like you are trying to optimize user experience for a very specific scenario while ignoring others who would expect the reconciliation to be paused, including the deletion.

Who are the others, i.e. what is the usecase we want to support?

We indicate that reconciliation is paused via condition of type `Synced` like this:

```yaml
reason: ReconcilePaused
status: "False"
type: Synced
```

when `False` it means that we are not syncing claim changes with the counterpart composite, i.e. we are not passing claim updates down the line. However, object deletion is not an update, at least not in CRUD/REST context, and therefore an user should be very surprised to discover that deletion of paused objects or higher order objects get stuck.

In order to [explore/learn](https://github.com/crossplane/crossplane/issues/4839) more about the use case mentioned by @turkenh I have updated the PR to match the behavior already available for claims/composites. Furthermore, I have added to the pause status condition a human readable message explain the consequences, and in the case of deletion we generate an additional k8s event and a log entry on info level.